vpn-proxy/app/models.py

```diff
@@ -6,58 +6,100 @@
-    """Find an available server IP in the given network (CIDR notation)"""
-    network = netaddr.IPNetwork(network)
-    if network.version != 4 or not network.is_private():
-        raise Exception("Only private IPv4 networks are supported.")
-    first, last = network.first, network.last
-    if first % 2:
-        first += 1
-    for i in range(20):
-        addr = str(netaddr.IPAddress(random.randrange(first, last, 2)))
-        print addr
-        try:
-            Tunnel.objects.get(server=addr)
-        except Tunnel.DoesNotExist:
-            return addr
-
-
-def check_server_ip(addr):
-    """Verify that the server IP is valid"""
-    addr = netaddr.IPAddress(addr)
-    if addr.version != 4 or not addr.is_private():
-        raise ValidationError("Only private IPv4 networks are supported.")
-    if not 0 < addr.words[-1] < 254 or addr.words[-1] % 2:
-        raise ValidationError("Server IP's last octet must be even in the "
-                              "range [2,254].")
-
-
-def check_destination_ip(addr):
-    """Verify that remote IP belongs to a private host"""
+def choose_server_ip(addr, exc_nets):
+    """Find an available server IP in the given network (CIDR notation)
+    based on the client IP supplied"""
+    address = netaddr.IPAddress(addr)
+    for network in ALLOWED_VPN_ADDRESSES:
+        if address in netaddr.IPNetwork(network):
+            # start iterating over the CIDR the IP belongs to
+            ALLOWED_VPN_ADDRESSES.remove(network)
+            ALLOWED_VPN_ADDRESSES.append(network)
+            break
+    else:
+        raise ValidationError('IP address is outside the supported range')
+    for network in reversed(ALLOWED_VPN_ADDRESSES):
+        cidr = netaddr.IPNetwork(network)
+        first, last = cidr.first, cidr.last
+        for exc_net in exc_nets:
+            if exc_net in cidr:
+                last -= (exc_net.last - exc_net.first + 1)
+        for _ in range(first + 1, last):
+            address += 1
+            for exc_net in exc_nets:
+                if address in exc_net:
+                    address = netaddr.IPAddress(exc_net.last + 1)
+            if address not in cidr or address == cidr.broadcast:
+                address = cidr.network + 1
+            try:
+                Tunnel.objects.get(Q(server=str(address)) |
+                                   Q(client=str(address)))
+            except Tunnel.DoesNotExist:
+                return str(address)
+
+
+def choose_client_ip(cidrs, excluded=[], reserved=EXCLUDED_VPN_ADDRESSES):
+    """Find an available client IP in one of the available private networks"""
+    exc_nets = []
+    for res_net in reserved:
+        exc_nets.append(netaddr.IPNetwork(res_net))
+    for exc_net in excluded:
+        exc_nets.append(netaddr.IPNetwork(exc_net))
+    for cidr in cidrs:
+        exc_nets.append(netaddr.IPNetwork(cidr))
+    # a list of unique, non-overlapping supernets
+    exc_nets = netaddr.IPSet(exc_nets).iter_cidrs()
```

Comment:
Instead of using nested loops, it is preferable to create a single `IpSet` containing the allowed addresses (`ALLOWED_VPN_ADDRESSES - cidrs - excluded - reserved`) and iterate over that. It will greatly simplify the code.